### PR TITLE
SpdxDocumentFile: Support transitive dependencies

### PR DESCRIPTION
**Pull request by oheger-bosch**
Refer to the single commits for further details.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                loader,
+                managerName,
+                rootDocument,
+                rootDocumentFile.toURI(),
+                references,
+                issues,
+                mutableSetOf()
+            )
+
+            val packages = collectPackages(references) + rootDocument.getPackages()
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:103]: The `+` potentially overwrites entries in the map. Is this intended? Should we use our own `zip(WithDefault)` functions instead? If there never could be a key-clash, please add a code comment why.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    fun getSpdxPackageForId(identifier: String, issues: MutableList<OrtIssue>): SpdxPackage? {
+        return packagesById[identifier]?.also {
+            issueForPackage(identifier)?.let { issues += it }
+        } ?: unresolvedPackage(identifier, issues)
+    }
+
+    /**
+     * Generate an issue for the package with [identifier] that cannot be resolved and add it to [issues].
+     */
+    private fun unresolvedPackage(identifier: String, issues: MutableList<OrtIssue>): SpdxPackage? {
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:136]: This function does not return / create an "unresolved package" like the name suggests, but always `null`. So at a minimum, the function should be renamed. Probably better yet, inline it. I'll come up with a proposal.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        }
+    }
+
+    /**
+     * Get the [SpdxPackage] for the given [identifier] by resolving against packages or external document references
+     * contained in this document. If the package cannot be resolved, add an issue to [issues].
+     */
+    fun getSpdxPackageForId(identifier: String, issues: MutableList<OrtIssue>): SpdxPackage? {
+        return packagesById[identifier]?.also {
+            issueForPackage(identifier)?.let { issues += it }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:115]: IMO the `let` should also be `also` instead as we're not interested in the return value.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    return allPackages
+}
+
+/**
+ * A data class to hold the result of an operation to resolve an [SpdxDocument] from an external reference. Resolving
+ * of the document may fail, then the document is *null*, and a corresponding [OrtIssue] is present.
+ */
+private data class DocumentResolveResult(
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:239]: Maybe "DocumentResolutionResult" or just "ResolutionResult"?

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+    loader: SpdxDocumentLoader,
+    baseUri: URI,
+    managerName: String
+): DocumentResolveResult {
+    val spdxFile = File(uri).absoluteFile.normalize()
+
+    return spdxFile.takeIf { it.isFile }?.let {
+        DocumentResolveResult(loader.load(it), uri, verifyChecksum(it, baseUri, managerName))
+    }
+        ?: DocumentResolveResult(
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:308]: Merge lines.

File: analyzer/src/test/kotlin/managers/SpdxTransitiveDocumentTest.kt [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+import org.ossreviewtoolkit.utils.spdx.model.SpdxExternalDocumentReference
+import org.ossreviewtoolkit.utils.spdx.model.SpdxPackage
+import org.ossreviewtoolkit.utils.spdx.model.SpdxRelationship
+import org.ossreviewtoolkit.utils.test.createTestTempDir
+import org.ossreviewtoolkit.utils.test.shouldNotBeNull
+
+class SpdxTransitiveDocumentTest : WordSpec() {
+    private lateinit var tempDir: File
+
+    /** A mock server used for testing whether external references can reference documents via the internet. */
```
Comment (sschuberth) [inline analyzer/src/test/kotlin/managers/SpdxTransitiveDocumentTest.kt:60]: "Internet"

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 7]
```diff
@@ -177,6 +187,18 @@ private fun collectPackages(
...
 }
 
+/**
+ * Return a list with all [SpdxRelationship]s found in the given [references]. Qualify the identifiers used in these
+ * relationships, so that they are compatible with the keys used to access the aggregated packages.
+ */
+private fun collectAndQualifyRelations(
+    references: MutableMap<SpdxExternalDocumentReference, SpdxDocument>
+): List<SpdxRelationship> {
+    return references.flatMap { (reference, document) ->
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt]: Could be a function expression.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+import org.ossreviewtoolkit.utils.spdx.model.SpdxExternalDocumentReference
+import org.ossreviewtoolkit.utils.spdx.model.SpdxPackage
+
+/**
+ * A data class storing information about a root SPDX document and all the documents referenced by it.
+ *
+ * This class is used by [SpdxDocumentFile] to get a combined view on all packages and their relations defined in a set
+ * of SPDX files.
+ */
+internal data class SpdxTransitiveDocument(
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:49]: Renamed the class to `SpdxResolvedDocument`.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt (hunk 1 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:103]: Added a comment.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:103]: BTW, I was in particular thinking about cyclic dependencies. Would this be that place to catch them?

File: analyzer/src/main/kotlin/managers/SpdxDocumentLoader.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * A helper class for [[SpdxDocumentFile]] that deals with loading of SPDX documents.
+ *
+ * While loading of SPDX documents is not that complicated, this class mainly is responsible for caching loaded
+ * documents to prevent that they are loaded multiple times - which could be the case for instance if multiple
+ * packages are resolved from an external document reference.
+ *
+ * Implementation note: This implementation is not thread-safe.
+ */
+internal class SpdxDocumentLoader {
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentLoader.kt:36]: Yet one more renaming suggestion: How about "SpdxDocumentCache"? This emphasizes that the main task is caching, not loading, and gets rid of the repetitive `loader.load()` (which would become `cache.load()`).

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt (hunk 2 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:136]: Changed.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt (hunk 4 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:239]: Renamed to `Resolutionresult`.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt [hunk 10]
```diff
@@ -219,6 +252,17 @@ private data class DocumentResolveResult(
     val issue: OrtIssue?
 )
 
+/**
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:255]: Not sure what you mean here. Which changes exactly?

[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:255]: The introduction and use of the extension functions, i.e. really all changes starting from line 255 until the end of the file. Or am I missing something and these changes are not as independent as they seem to be?

File: analyzer/src/main/kotlin/managers/SpdxDocumentLoader.kt (hunk 9 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxDocumentLoader.kt:36]: Renamed to `SpdxDocumentCache`.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt (hunk 1 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:103]: No, cycles in the dependencies are handled by `SpdxDocumentFile` when processing the relationships. `SpdxResolvedDocument` detects cycles in references in the `resolveAllReferences()` function.

File: analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt:255]: I may be missing something, but these functions are part of the initial implementation and used directly. So splitting these into a separate commit would mean a high effort, as the implementation would need to be changed, too.

[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt]: Sorry, I probably got confused by the (temporary) copying of code from `SpdxDocumentFile`.

File: analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  name: "my-lib"
+  versionInfo: "8.88.8"
+  originator: "Person: Harry Hirsch (hirsch@forest.org)"
+relationships:
+- spdxElementId: "DocumentRef-curl-7.70.0:SPDXRef-Package-curl"
+  relationshipType: "DEPENDENCY_OF"
+  relatedSpdxElement: "SPDXRef-Package-my-lib"
+- spdxElementId: "DocumentRef-openssl-1.1.1g:SPDXRef-Package-openssl"
+  relationshipType: "DEPENDENCY_OF"
+  relatedSpdxElement: "SPDXRef-Package-my-lib"
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml:52]: For the fun of testing, should we swap source and target here and use `DEPENDS_ON`?

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt [hunk 12]
```diff
@@ -403,97 +402,62 @@ class SpdxDocumentFile(
...
-        doc: SpdxDocument,
-        definitionFile: File,
+        pkgId: String,
+        doc: SpdxResolvedDocument,
         packages: MutableSet<Package>,
         dependencyOfRelation: SpdxRelationship.Type,
         dependsOnCase: (String) -> PackageReference? = { null }
     ): SortedSet<PackageReference> =
         doc.relationships.mapNotNullTo(sortedSetOf()) { (source, relation, target) ->
+            println("$pkgId, $dependencyOfRelation: ($source) --$relation--> ($target)")
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:335]: Debugging left-over.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt [hunk 13]
```diff
@@ -22,8 +22,6 @@
 package org.ossreviewtoolkit.analyzer.managers
 
 import java.io.File
-import java.net.Authenticator
-import java.net.URI
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:26]: The commit message still refers to `SpdxTransitiveDocument`.

File: analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml [hunk 14]
```diff
@@ -0,0 +1,55 @@
+SPDXID: "SPDXRef-DOCUMENT"
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml:1]: The commit message still refers to `SpdxTransitiveDocument`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]

File: analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml (hunk 11 above)
Comment (JSurf) [reply analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml:52]: Actually i just stumbled upon this while trying to figure out why my transitive dependencies do not show up. So is "DEPENDS_ON" broken ? Switching to DEPENDENCY_OF made them show up

Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml:52]: With respect to ORT's SPDX document file parsing, I believe both relations should be supported to express dependencies in one direction or the other, also see

https://github.com/oss-review-toolkit/ort/blob/3355b85803936a01ba706fffbad9ab3978b44ffe/plugins/package-managers/spdx/src/main/kotlin/SpdxDocumentFile.kt#L395-L406

Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/synthetic/spdx/libs/my-lib/package.spdx.yml:52]: As it's a bit hard to follow this discussion on a closed PR, I've created https://github.com/oss-review-toolkit/ort/issues/8359. Please continue the discussion in there, @JSurf.
